src/Constraint/MultiConstraint.php

```diff
@@ -249,6 +247,29 @@ private static function optimizeConstraints(array $constraints, $conjunctive)
                     && $right1[0] === '<'
                     && substr($left1, 2) === substr($right0, 3)
                 ) {
+                    for ($j = 2, $m = count($right->constraints); $j < $m; $j++) {
+                        $c = (string)$right->constraints[$j];
+                        switch ($c[0]) {
+                            case '!':
+                            case '<':
+                            case '=':
+                                break;
+                            default:
+                                continue 3;
+                        }
+                    }
+                    for ($j = 2, $m = count($left->constraints); $j < $m; $j++) {
+                        $c = (string)$left->constraints[$j];
+                        switch ($c[0]) {
+                            case '!':
+                            case '>':
+                            case '=':
+                                break;
+                            default:
+                                continue 3;
+                        }
+                    }
```

Comment:
Not against doing this, but I think some comments would be good here :)

Why !/>/= and !/</= are OK and the rest not?

I see how != is fine to optimize as that's a single gap in the range. 

But I don't see why == is ok to optimize, `^1.0 || ^2.0, =2.0.3` means `^1.0 || 2.0.3` but with this optimization it'd become `1.0 - 3.0, 2.0.3` which is only matched by `2.0.3`. The `^1.0` is dropped.

I am also not sure why </> are ok in each respective loop. `^1.0 <1.1 || ^2.0 > 2.0.4` getting merged to `1.0 - 3.0, <1.1, > 2.0.4` also seems wrong to me as this suddenly cannot match anything anymore.

So all in all this seems broken, except for != which can be safely kept, but then the question is it worth looping over things just in the odd case we have a !=, which is anyway quite rare..

Reply:
yeah.. you're right, I missed to much cases... let's close this